controllers/events/eventsource_controller.go

```diff
@@ -0,0 +1,476 @@
+		} else {
+			eventBusSpec = clusterEventBus.Spec
+		}
+	} else {
+		eventBusSpec = eventBus.Spec
+	}
+
+	componentName := fmt.Sprintf(EventSourceBusComponentNameTmpl, eventSource.Name)
+
+	// Set SourceConfig.EventBusComponentName.
+	r.envs.EventBusComponentName = componentName
+
+	// Generate a dapr component based on the specification of EventBus.
+	component := &componentsv1alpha1.Component{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      componentName,
+			Namespace: eventSource.Namespace,
+		},
+	}
+
+	if eventBusSpec.Nats != nil {
+		specBytes, err := json.Marshal(eventBusSpec.Nats)
+		if err != nil {
+			return err
+		}
+
+		// Set the SourceConfig.EventBusSpecEncode which will reflect the specification content changes of dapr component back to the EventSource workload,
+		// informing it that it needs to rebuild.
+		r.envs.EventBusSpecEncode = base64.StdEncoding.EncodeToString(specBytes)
+
+		// Create the dapr component for EventSource to send event to EventBus.
+		if _, err := ctrl.CreateOrUpdate(r.ctx, r.Client, component, mutateDaprComponent(r.Scheme, component, eventSource, eventBusSpec.Nats)); err != nil {
+			return err
+		}
+		return nil
+	}
+
+	return errors.New("no specification found for create dapr component")
+}
+
+func (r *EventSourceReconciler) handleSink(eventSource *openfunctionevent.EventSource) error {
+	componentName := fmt.Sprintf(EventSourceSinkComponentNameTmpl, eventSource.Name)
+
+	// Set SourceConfig.SinkComponentName.
+	r.envs.SinkComponentName = componentName
+
+	// Generate a dapr component based on the specification of Sink.
+	component := &componentsv1alpha1.Component{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      componentName,
+			Namespace: eventSource.Namespace,
+		},
+	}
+
+	spec, err := newSinkComponentSpec(r.Client, r.Log, eventSource.Spec.Sink.Ref)
+	if err != nil {
+		return err
+	}
+	specBytes, err := json.Marshal(spec)
+	if err != nil {
+		return err
+	}
+
+	// Set the SourceConfig.SinkSpecEncode which will reflect the specification content changes of dapr component back to the EventSource workload,
+	// informing it that it needs to rebuild.
+	r.envs.SinkSpecEncode = base64.StdEncoding.EncodeToString(specBytes)
+
+	// Create the dapr component for EventSource to retrieve event from EventBus.
+	if _, err := ctrl.CreateOrUpdate(r.ctx, r.Client, component, mutateDaprComponent(r.Scheme, component, eventSource, spec)); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (r *EventSourceReconciler) handleEventSource(eventSource *openfunctionevent.EventSource) error {
+	type sourceSpec struct {
+		*componentsv1alpha1.Component
+		SourceTopic string
+		SourceKind  string
+		EventName   string
+	}
+	var sourceSpecs []*sourceSpec
+
+	// Generate dapr components based on the specification of EventSource.
+	if eventSource.Spec.Kafka != nil {
+		for eventName, spec := range eventSource.Spec.Kafka {
+			ss := &sourceSpec{}
+			componentSpec := spec.ComponentSpec
```

Comment:
It's better to set the consumerGroup of kafka to a unique name such as the component namespace name